### PR TITLE
Block, VM: Unify hardforkBy Options

### PR DESCRIPTION
**Pull request by holgerd77**
Second part of #2774 following #2798

This PR does the following:

- Block: unify `hardforkByBlockNumber`, `hardforkByTTD` options -> `setHardfork`
- VM (constructor/runBlock()): unify `hardforkByBlockNumber`, `hardforkByTTD` options -> `setHardfork`

### CODE REVIEW

**Review by jochem-brouwer — DISMISSED** [review]
LGTM

File: packages/block/src/header.ts [hunk 1]
```diff
@@ -182,11 +177,17 @@ export class BlockHeader {
...
-    const hardforkByBlockNumber = options.hardforkByBlockNumber ?? false
-    if (hardforkByBlockNumber || options.hardforkByTTD !== undefined) {
+    const setHardfork = options.setHardfork ?? false
+    if (setHardfork === true) {
+      this._common.setHardforkBy({
+        blockNumber: number,
+        timestamp,
+      })
+    }
+    if (typeof setHardfork !== 'boolean') {
```
Comment (acolytec3) [inline packages/block/src/header.ts:187]: Could this be simplified to just the below?

```suggestion
    else {
```

Comment (acolytec3) [inline packages/block/src/header.ts:187]:
```suggestion
    else if (typeof setHardfork !== 'boolean') {
```

File: packages/vm/src/runBlock.ts [hunk 2]
```diff
@@ -56,16 +57,21 @@ export async function runBlock(this: VM, opts: RunBlockOpts): Promise<RunBlockRe
...
-    })
+  if (setHardfork !== false || this._setHardfork !== false) {
+    const setHardforkUsed = setHardfork ?? this._setHardfork
+    if (setHardforkUsed === true) {
+      this._common.setHardforkBy({
+        blockNumber: block.header.number,
+        timestamp: block.header.timestamp,
+      })
+    }
+    if (typeof setHardforkUsed !== 'boolean') {
```
Comment (acolytec3) [inline packages/vm/src/runBlock.ts:68]:
```suggestion
    else if (typeof setHardforkUsed !== 'boolean') {
```

**Review by acolytec3 — CHANGES_REQUESTED** [review]
I could be missing the significance but requested one change in how we evaluate the `setHardfork` opt so we don't check it twice (if it's indeed `true`).  Small efficiency gain if I'm not missing something.  Otherwise, looks great!

**Comment by holgerd77** [comment]
@jochem-brouwer @acolytec3 thanks for reviewing 🙂, have applied the suggested changes, needs re-approval now! 🙂

**Review by jochem-brouwer — DISMISSED** [review]
Ah right, good catch @acolytec3. LGTM still! :smile: (assuming CI passes)

**Comment by jochem-brouwer** [comment]
Nvm, CI fails. I should not approve so fast :sweat_smile: (how do I dismiss my review?)

**Comment by jochem-brouwer** [comment]
Will re-approve if CI passes

**Review by g11tech — APPROVED** [review]
looks amazing! :rocket: :heart:

**Pull request merged**